File: app/src/main/java/com/github/overpass/gather/screen/base/personal/DataFragment.java
package com.github.overpass.gather.screen.base.personal;

import android.content.Intent;
import android.net.Uri;
import android.view.View;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.airbnb.lottie.LottieAnimationView;
import com.github.overpass.gather.R;
import com.github.overpass.gather.screen.auth.register.RegistrationFragment;
import com.github.overpass.gather.screen.auth.register.add.AddDataStatus;
import com.github.overpass.gather.screen.auth.register.add.ImageSource;
import com.github.overpass.gather.screen.dialog.PickImageDialogFragment;
import com.github.overpass.gather.screen.dialog.progress.indeterminate.ProgressDialogFragment;
import com.google.android.material.textfield.TextInputEditText;

import butterknife.BindView;
import butterknife.OnLongClick;

import static com.github.overpass.gather.model.commons.UIUtil.snackbar;
import static com.github.overpass.gather.model.commons.UIUtil.textOf;

// TODO: The hierarchy is messed up
public abstract class DataFragment<VM extends DataViewModel, C>
        extends RegistrationFragment<VM, C> {

    @Nullable
    @BindView(R.id.tietUsername)
    protected TextInputEditText tietUsername;
    @BindView(R.id.ivPhotoPreview)
    protected ImageView ivPhotoPreview;
    @Nullable
    @BindView(R.id.lavTick)
    protected LottieAnimationView lavTick;

    @Override
    protected void onBind() {
        super.onBind();
        getViewModel().getChosenImageData().observe(getViewLifecycleOwner(), this::handleChosenImageUri);
        getViewModel().getImageSourceData().observe(getViewLifecycleOwner(), this::handleImageSource);
        getViewModel().getWritePermissionDeniedData()
                .observe(getViewLifecycleOwner(), this::onPermissionChanged);
        getViewModel().getReadPermissionDeniedData()
                .observe(getViewLifecycleOwner(), this::onPermissionChanged);
    }

    protected void onPermissionChanged(boolean denied) {
        if (denied) {
            snackbar(ivPhotoPreview, "Sorry, you can't choose photo without permissions");
        }
    }

    protected void handleImageSource(ImageSource imageSource) {
        if (imageSource == ImageSource.GALLERY) {
            getViewModel().chooseFromGallery(getActivity(), this);
        } else if (imageSource == ImageSource.CAMERA) {
            getViewModel().chooseFromCamera(getActivity(), this);
        }
    }

    protected void handleChosenImageUri(@Nullable Uri uri) {
        if (uri != null) {
            ivPhotoPreview.setScaleX(1);
            ivPhotoPreview.setScaleY(1);
            ivPhotoPreview.setScaleType(ImageView.ScaleType.CENTER_CROP);
            ivPhotoPreview.setImageURI(uri);
            lavTick.setVisibility(View.VISIBLE);
            lavTick.playAnimation();
        } else { // reset
            ivPhotoPreview.setScaleY(2);
            ivPhotoPreview.setScaleX(2);
            ivPhotoPreview.setScaleType(ImageView.ScaleType.CENTER);
            ivPhotoPreview.setImageResource(R.drawable.ic_add_pic);
            lavTick.setVisibility(View.GONE);
        }
    }

    protected void onChooseImageClick() {
        PickImageDialogFragment.show(getFragmentManager());
    }

    @OnLongClick(R.id.ivPhotoPreview)
    protected void resetImage() {
        getViewModel().resetChosenImage();
    }

    public void onSubmitClick() {
        getViewModel().submit(getContext().getContentResolver(), textOf(tietUsername))
                .observe(getViewLifecycleOwner(), this::handleAddDataStatus);
    }

    private void handleAddDataStatus(AddDataStatus addDataStatus) {
        switch (addDataStatus.tag()) {
            case AddDataStatus.PROGRESS:
                handleProgress(addDataStatus.as(AddDataStatus.Progress.class));
                break;
            case AddDataStatus.INVALID_USERNAME:
                handleInvalidUsername(addDataStatus.as(AddDataStatus.InvalidUsername.class));
                break;
            case AddDataStatus.ERROR:
                handleError(addDataStatus.as(AddDataStatus.Error.class));
                break;
            case AddDataStatus.SUCCESS:
                handleSuccess(addDataStatus.as(AddDataStatus.Success.class));
                break;
        }
    }

    protected void handleProgress(AddDataStatus.Progress progress) {
        ProgressDialogFragment.Companion.show(getFragmentManager());
    }

    protected void handleInvalidUsername(AddDataStatus.InvalidUsername invalidUsername) {
        ProgressDialogFragment.Companion.hide(getFragmentManager());
        tietUsername.setError("Invalid username");
    }

    protected void handleError(AddDataStatus.Error error) {
        ProgressDialogFragment.Companion.hide(getFragmentManager());
        String message = error.getThrowable().getLocalizedMessage();
        snackbar(ivPhotoPreview, message);
    }

    protected void handleSuccess(AddDataStatus.Success success) {
        ProgressDialogFragment.Companion.hide(getFragmentManager());
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        getViewModel().onImageChosen(requestCode, resultCode, data);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        getViewModel().onRequestPermissionsResult(requestCode, permissions, grantResults, this);
    }
}

File: app/src/main/java/by/overpass/gather/ui/base/personal/DataFragment.kt
package by.overpass.gather.ui.base.personal

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import androidx.lifecycle.Observer
import by.overpass.gather.R
import by.overpass.gather.commons.android.lifecycle.on
import by.overpass.gather.commons.android.lifecycle.onMaybeNull
import by.overpass.gather.commons.android.snackbar
import by.overpass.gather.commons.android.text
import by.overpass.gather.ui.auth.register.RegistrationFragment
import by.overpass.gather.ui.auth.register.add.AddDataStatus
import by.overpass.gather.ui.auth.register.add.ImageSource
import by.overpass.gather.ui.dialog.PickImageDialogFragment
import by.overpass.gather.ui.dialog.progress.indeterminate.ProgressDialogFragment
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.synthetic.main.content_personal_data.*

// TODO: The hierarchy is messed up
abstract class DataFragment<VM : DataViewModel> : RegistrationFragment<VM>() {

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        ivPhotoPreview.setOnLongClickListener {
            viewModel.resetChosenImage()
            true
        }
    }

    override fun onBind() {
        super.onBind()
        onMaybeNull(viewModel.getChosenImageData()) {
            handleChosenImageUri(it)
        }
        on(viewModel.imageSourceData) {
            handleImageSource(it)
        }
        on(viewModel.getWritePermissionDeniedData()) {
            onPermissionChanged(it)
        }
        on(viewModel.getReadPermissionDeniedData()) {
            onPermissionChanged(it)
        }
    }

    protected fun onPermissionChanged(denied: Boolean) {
        if (denied) {
            ivPhotoPreview!!.snackbar("Sorry, you can't choose photo without permissions", Snackbar.LENGTH_SHORT)
        }
    }

    protected fun handleImageSource(imageSource: ImageSource) {
        if (imageSource == ImageSource.GALLERY) {
            viewModel.chooseFromGallery(requireActivity(), this)
        } else if (imageSource == ImageSource.CAMERA) {
            viewModel.chooseFromCamera(requireActivity(), this)
        }
    }

    protected open fun handleChosenImageUri(uri: Uri?) {
        if (uri != null) {
            ivPhotoPreview.scaleX = 1f
            ivPhotoPreview.scaleY = 1f
            ivPhotoPreview.scaleType = ImageView.ScaleType.CENTER_CROP
            ivPhotoPreview.setImageURI(uri)
            lavTick.visibility = View.VISIBLE
            lavTick.playAnimation()
        } else { // reset
            ivPhotoPreview.scaleY = 2f
            ivPhotoPreview.scaleX = 2f
            ivPhotoPreview.scaleType = ImageView.ScaleType.CENTER
            ivPhotoPreview.setImageResource(R.drawable.ic_add_pic)
            lavTick.visibility = View.GONE
        }
    }

    protected open fun onChooseImageClick() {
        PickImageDialogFragment.show(fragmentManager)
    }

    fun onSubmitClick() {
        viewModel.submit(context!!.contentResolver, tietUsername.text())
                .observe(viewLifecycleOwner, Observer<AddDataStatus> { this.handleAddDataStatus(it) })
    }

    private fun handleAddDataStatus(addDataStatus: AddDataStatus) {
        when (addDataStatus.tag()) {
            AddDataStatus.PROGRESS -> handleProgress(addDataStatus.`as`<AddDataStatus.Progress>(AddDataStatus.Progress::class.java))
            AddDataStatus.INVALID_USERNAME -> handleInvalidUsername(addDataStatus.`as`<AddDataStatus.InvalidUsername>(AddDataStatus.InvalidUsername::class.java))
            AddDataStatus.ERROR -> handleError(addDataStatus.`as`<AddDataStatus.Error>(AddDataStatus.Error::class.java))
            AddDataStatus.SUCCESS -> handleSuccess(addDataStatus.`as`<AddDataStatus.Success>(AddDataStatus.Success::class.java))
        }
    }

    protected fun handleProgress(progress: AddDataStatus.Progress) {
        ProgressDialogFragment.show(fragmentManager)
    }

    protected fun handleInvalidUsername(invalidUsername: AddDataStatus.InvalidUsername) {
        ProgressDialogFragment.hide(fragmentManager)
        tietUsername.error = "Invalid username"
    }

    protected fun handleError(error: AddDataStatus.Error) {
        ProgressDialogFragment.hide(fragmentManager)
        val message = error.throwable.localizedMessage
        ivPhotoPreview.snackbar(message, Snackbar.LENGTH_SHORT)
    }

    protected open fun handleSuccess(success: AddDataStatus.Success) {
        ProgressDialogFragment.hide(fragmentManager)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        viewModel.onImageChosen(requestCode, resultCode, data)
    }

    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>,
                                            grantResults: IntArray) {
        viewModel.onRequestPermissionsResult(requestCode, permissions, grantResults, this)
    }
}
